### PR TITLE
Add a tidied up version of the JWT token stuff from ViaHTML security POC

### PR DESCRIPTION
**Pull request by jon-betts**
This adds:

 * A basic wrappers around a JWT based tokens
 * A more specialised one for encoding and comparing paths
 * One specifically for doing this in a ViaHTML context

## Review notes

 * Nothing here is used yet, but it will be used in ViaHTML at least in: https://github.com/hypothesis/viahtml/pull/47
 * Actual usage can be seen in: https://github.com/hypothesis/viahtml/blob/lockdown-poc/viahtml/views/authentication.py
 * This is somewhat usage agnostic. Most of this could live in a totally separate lib
 * It might be a good thing to carve off into `checkmate-lib`
 * Some of the things we can do, we don't happen to yet. But this lets us experiment and change our minds. For example we mostly use exact expiry times, not not 100% of the time. We also don't pass timedeltas for max age, but passing `timedelta(hour=2)` is a lot nicer than `60 * 60 * 2`.

Here's a small script which uses this stuff to sign ViaHTML urls:

```python
from h_vialib.configuration import Configuration
from h_vialib.secure import ViaSecureURLToken


def sign_url(url, secret, via_host):
    sec = ViaSecureURLToken(secret).create(url)

    signed_url = Configuration.add_to_url(
        url, via_params={"sec": sec},
        client_params={})

    return f"{via_host}/{signed_url}"


url = "http://example.com?a=ga"
print(sign_url(url, secret="not_a_secret", via_host="http://localhost:9085/proxy"))
```

### CODE REVIEW

File: src/h_vialib/secure/token.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+        return urlunparse(parts)
+
+
+class ViaSecureURLToken(SecureURLToken):
+    """A token for signing proxied URLs."""
+
+    MAX_AGE = 60 * 60  # An hour
+
+    def create(self, proxied_url):  # pylint: disable=arguments-differ
```
Comment (marcospri) [inline src/h_vialib/secure/token.py:133]: Doesn't this need a `secret` param or similar?

Comment (jon-betts) [reply src/h_vialib/secure/token.py:133]: It has a secret by virtue of the parent class

File: src/h_vialib/secure/expiry.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    # Timezone aware dates are annoying, but allow you to include the TZ in a
+    # serialisation, which is required for cookie expiry times
+    delta = datetime.now(tz=timezone.utc) - YEAR_ZERO
+
+    quantized_expire = (delta.total_seconds() // quantization) * quantization
+
+    return YEAR_ZERO + timedelta(seconds=quantized_expire + max_age)
+
+
+def as_expires(expires=None, max_age=None):
```
Comment (jon-betts) [inline src/h_vialib/secure/expiry.py:46]: I split this off so anything which has this pattern (like most of our secure things) can have all this fancy conversion and parsing without having to code for it everywhere

File: src/h_vialib/secure/token.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        except InvalidSignatureError as err:
+            raise InvalidToken("Invalid secure token") from err
+        except ExpiredSignatureError as err:
+            raise InvalidToken("Expired secure token") from err
+        except DecodeError as err:
+            raise InvalidToken("Malformed secure token") from err
+
+
+class SecureURLToken(SecureToken):
+    """A secure token used for signing and checking URLs."""
```
Comment (jon-betts) [inline src/h_vialib/secure/token.py:62]: We don't directly use this inbetween class, so the two could be merged, _but..._ there's nothing Via specific up until this level. If we wanted to put this in another general lib, this is where the split would be.

File: src/h_vialib/secure/token.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        if not parts.path:
+            parts = parts._replace(path="/")
+
+        return urlunparse(parts)
+
+
+class ViaSecureURLToken(SecureURLToken):
+    """A token for signing proxied URLs."""
+
+    MAX_AGE = 60 * 60  # An hour
```
Comment (marcospri) [inline src/h_vialib/secure/token.py:131]: I'd personally prefer this to be an optional parameter with a default in `create`.

File: src/h_vialib/secure/token.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+        # Ensure that http://example.com and http://example.com/ are considered
+        # the same
+        if not parts.path:
+            parts = parts._replace(path="/")
+
+        return urlunparse(parts)
+
+
+class ViaSecureURLToken(SecureURLToken):
```
Comment (marcospri) [inline src/h_vialib/secure/token.py:128]: I'm not sure if this is necessary: https://github.com/hypothesis/via3/pull/357/files#diff-05da9214f2b26c51ec3ce88c7ea599b06800f7ced042160d1cf1be94dec7b4c6R55

if it is, would it make sense to have overload `verify` here to accept the full URL and remove the query param?

File: tests/unit/h_vialib/secure/token_test.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    def test_create_works(self, token):
+        token_string = token.create("http://example.com?via.config=blah")
+
+        assert token_string == Any.string()
+        assert decode_token(token_string) == {
+            "url": "http://example.com/",
+            "exp": Any.int(),
+        }
+
+    def test_create_uses_a_quantized_expiry(self, token, quantized_expiry):
```
Comment (marcospri) [inline tests/unit/h_vialib/secure/token_test.py:135]: This tests fails for me locally, shouldn't `quantized_expiry.return_value` be a bit in the future (vs just `now()`)?

File: src/h_vialib/secure/token.py (hunk 5 above)
Comment (jon-betts) [reply src/h_vialib/secure/token.py:128]: It isn't necessary _as long as the param is prefixed with `via.`_. `ViaSecureURLToken` removes all Via config prior to comparison in it's normalisation function.

File: tests/unit/h_vialib/secure/token_test.py (hunk 6 above)
Comment (jon-betts) [reply tests/unit/h_vialib/secure/token_test.py:135]: What failure are you getting?

Comment (marcospri) [reply tests/unit/h_vialib/secure/token_test.py:135]: Sorry, looks like:
```AssertionError: assert {'exp': 1612787391, 'url': '*any*'} == <AnyDict '* any dict containing {'exp': 1612783791} *'>
E        +  where {'exp': 1612787391, 'url': '*any*'} = decode_token('eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1cmwiOiIqYW55KiIsImV4cCI6MTYxMjc4NzM5MX0.8nzPoBo9rz8xf0_mPgwGXqnpseWWvY889hpPNttmsFg')
E        +  and   <AnyDict '* any dict containing {'exp': 1612783791} *'> = <function fluent_entrypoint.__get__.<locals>.wrapper at 0x7f8bcf087d08>({'exp': 1612783791})
E        +    where <function fluent_entrypoint.__get__.<locals>.wrapper at 0x7f8bcf087d08> = <class 'h_matchers.matcher.collection.AnyDict'>.containing
E        +      where <class 'h_matchers.matcher.collection.AnyDict'> = Any.dict
```

Comment (jon-betts) [reply tests/unit/h_vialib/secure/token_test.py:135]: I don't think the future / past makes a difference here, as we aren't checking that the token is valid, just that it has the contents we expect. It might be nice from a demonstrating the behavior point of view.

I'll scrub out my local tox envs and give this another shot. The core guts of the things being compared look identical, but I'm not sure about all the trappings.

We should go through `h-matchers` at some point, as it's our own home grown special testing sauce. It's very handy some times, but hard to understand and occasionally makes debugging tests a pain.

Comment (jon-betts) [reply tests/unit/h_vialib/secure/token_test.py:135]: Also missing the elephant in the room, it's passing in CI too

File: src/h_vialib/secure/token.py (hunk 4 above)
Comment (jon-betts) [reply src/h_vialib/secure/token.py:131]: I'm not sure about this. In the end we intend for this to be used by multiple products (maybe not directly, but once wrapped in whatever URL generation code we provide).

I don't think this is something we want to choose in multiple places, or have be different from product to product. It's not really variable.

File: src/h_vialib/exceptions.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+class InvalidToken(Exception):
+    """A security check has failed on a token."""
+
+
+class MissingToken(InvalidToken):
+    """An expected token is missing."""
+
+    # The missing token is a child of InvalidToken so you can distinguish
+    # between the two, or catch them together as you see fit
```
Comment (seanh) [inline src/h_vialib/exceptions.py]: Actually seems like it belongs on the docstring

File: src/h_vialib/secure/expiry.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    """Create a quantized expiry time.
+
+    The time returned is guaranteed be:
+        * A multiple of `max_age / divisions`
+        * At most `max_age` seconds in the future
+        * At least `max_age * (divisions - 1) / divisions` seconds in the future
+
+    :param max_age: The maximum age to issue (int, or timedelta)
+    :param divisions: How many subdivisions of the max age for quantization
+    :return: A datetime object
```
Comment (seanh) [inline src/h_vialib/secure/expiry.py]: I think the key thing about this is that if you call `quantized_expiry()` repeatedly with the same arguments it will keep on returning the same value until some amount of time (depending on `max_age` and `divisions`) has passed, then it'll start returning a different value, right? Rather than returning a different value every time you call it a millisecond later.

The seems like the key guarantee here / the reason this function exists so worth explaining in the docstring

File: src/h_vialib/secure/expiry.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    quantized_expire = (delta.total_seconds() // quantization) * quantization
+
+    return YEAR_ZERO + timedelta(seconds=quantized_expire + max_age)
+
+
+def as_expires(expires=None, max_age=None):
+    """Convert either an expiry time or max age to an expiry time.
+
+    :param expires: Datetime by which this token with expire
```
Comment (seanh) [inline src/h_vialib/secure/expiry.py]:
```suggestion
    :param expires: Datetime by which this token will expire
```

File: tests/unit/h_vialib/secure/expiry_test.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        assert diff > timedelta(seconds=0)
+        # It's at least max_age * divisions - 1 / divisions
+        assert diff >= timedelta(seconds=120 * (3 / 4))
+        # It's at most max_age
+        assert diff <= timedelta(seconds=120)
+
+    @pytest.mark.parametrize("max_age", (None, "foo"))
+    def test_it_raises_with_invalid_max_age(self, max_age):
+        with pytest.raises(ValueError):
+            quantized_expiry(max_age=max_age)
```
Comment (seanh) [inline tests/unit/h_vialib/secure/expiry_test.py]: Can we test that repeatedly calling it with the same arguments returns the same result? And then after an amount  of time has passed, it starts returning a different result? Similar to the tests I wrote here: https://github.com/hypothesis/via3/blob/c18e328fd946cf5411de893fc05d1b00d33857ac/tests/unit/via/views/view_pdf_test.py#L28-L69

File: src/h_vialib/secure/__init__.py [hunk 11]
```diff
@@ -0,0 +1,4 @@
+"""Security helpers."""
+
+from h_vialib.secure.expiry import quantized_expiry
+from h_vialib.secure.token import SecureToken, SecureURLToken, ViaSecureURLToken
```
Comment (seanh) [inline src/h_vialib/secure/__init__.py:4]: Can we add this stuff to `h-vialib/__init__.py`, like we do with `Configuration`: <https://github.com/hypothesis/h-vialib/blob/master/src/h_vialib/__init__.py>?

Reason: flat is better than nested. I just want to import what I need from `h_vialib`. I don't want to have to remember what comes from `h_vialib.configuration`, what comes from `h_vialib.secure`, etc. Instead of:

```python
from h_vialib.configuration import Configuration
from h_vialib.secure import ViaSecureURLToken
```

It could be:

```python
from h_vialib import Configuration, ViaSecureURLToken
```

**Review by seanh — CHANGES_REQUESTED** [review]
My main piece of feedback is that from the docstring and tests I think it's unclear why `quantized_expiry()` exists. We have a need to create expiry times where the time is a multiple of `max_age` and `divisions` and at most `max_age` seconds in the future and at least `max_age * (divisions - 1) / divisions` in the future? I mean, yes, but really we need this because we need to repeatedly create expiry times without them changing too frequently so that we don't destroy caching. I think we should clarify that in the docstring, and also test it

File: src/h_vialib/exceptions.py (hunk 7 above)
Comment (jon-betts) [reply src/h_vialib/exceptions.py]: Done

File: src/h_vialib/secure/expiry.py (hunk 8 above)
Comment (jon-betts) [reply src/h_vialib/secure/expiry.py]: I've added some more of the why, rather than just the what

File: tests/unit/h_vialib/secure/expiry_test.py (hunk 10 above)
Comment (jon-betts) [reply tests/unit/h_vialib/secure/expiry_test.py]: I think the conditions here actually are sufficient to prove the behavior in a mathematical sense, but I get that they don't show it intuitively. I'll add some tests which are easier to grok

File: src/h_vialib/secure/expiry.py (hunk 8 above)
Comment (seanh) [reply src/h_vialib/secure/expiry.py]: Should also add this to `h_vialib/__init__.py` since at least Via 3 is gonna be using it (just to make it more convenient to import)

File: src/h_vialib/secure/__init__.py (hunk 11 above)
Comment (jon-betts) [reply src/h_vialib/secure/__init__.py:4]: Maybe.  There will pretty soon be about 8 things in `__init__.py`, we might not want to put everything there for the same reason we don't keep all files in the root of our projects. Flat is better than nested except when it's the same or worse.

I'd also suggest that there are two distinct customers for `h-vialib`'s public interface:

* Actual types of Via product
* Client types (LMS, h, client?)

As a client, you maybe shouldn't be poking about into nested directories, despite them being "public". This gives us a way to hint at that. This also means we trade one type of confusion "where do I import things from?" for another "how many of these classes am I supposed to use as a client?" The difference being that Via type users are high information users, and can be expected to know. 

It's worth keeping both of these "problems" in proportion though. You need to "remember" which to use or where to get it from for the short period of time, and as long as it's not hard to find out, then git will do the remembering for you afterwards (Do you remember exactly where `HTTPFound` comes from in `pyramid`? Maybe maybe not, but it's not the root, and it's not hard to find out). I'd also recommend looking into hooking up some autocomplete in your IDE. I press "Alt + Enter" on a class name and Pycharm does a reasonable job of importing it for me _most_ of the time.

As we ship in and evolve the interface, we might want move the other way and hide `Configuration` from the root, as clients shouldn't need it, so there might only be two or three classes.

**Comment by jon-betts** [comment]
Changes from before:

 * Ensuring that when we convert a "max_age" to an expiry it has a timezone (required for cookies, I think)
 * Added more discussion of why you'd want a quantized expiry
 * Added more scenario based tests of the quantization
 
Things I've not changed:

 * The secure stuff is still in it's own directory
 * I think we have two different customers for the code: clients and Via implementers
 * I think it's reasonable for Via implementers to have knowledge of the layout to keep it simple for clients (who don't have to pick through a lot in the root)

File: src/h_vialib/exceptions.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+class InvalidToken(Exception):
+    """A security check has failed on a token."""
+
+
+class MissingToken(InvalidToken):
+    """An expected token is missing.
+
+    The missing token is a child of `InvalidToken` so you can distinguish
+    between the two, or catch them together as you see fit
+    """
```
Comment (seanh) [inline src/h_vialib/exceptions.py]: I think this makes it easy to write nonsensical code like:

```python
try:
    ...
except MissingToken:
    ...
except InvalidToken:
    ...
```

The code is trying to handle missing and invalid tokens differently but the `except InvalidToken` will never be hit because `MissingToken` is an `InvalidToken` subclass. If the two `except`'s were in the other order it'd work. I'm not sure whether PyLint or Python itself would catch this mistake.

Having a separate `TokenException` base class that both `InvalidToken` and `MissingToken` inherit from would mean that you can catch `InvalidToken` and `MissingToken` separately and it doesn't matter what order you put the `except`'s in, and if you _want_ to catch both with a single `except` you'd catch the `TokenException` base class instead.

Our standard guide for laying out exception classes would achieve this "automatically" (without even having to think of the `TokenException` base class) because all `vialib` exceptions would inherit from `ViaLibException` and all `vialib.secure` exceptions would inherit from `SecureException`. So you could catch `InvalidToken` and `MissingToken` separately in either order, or catch either `SecureException` or `ViaLibException` if you want to catch both. It'd also put `InvalidToken` and `MissingToken` closer to where they're used, in `secure/exceptions.py`. The separate `TokenException` _can_ still be added but isn't necessary.

```python
# h_vialib/exceptions.py

class ViaLibException(Exception):
    """Base class for all h_vialid exceptions."""
```

```python
# h_vialib/secure/exceptions.py

from h_vialib.exceptions import ViaLibException

class SecureException(ViaLibException):
    """Base class for h_vialib.secure exceptions."""

# You wouldn't really need this one anymore, but you could still do it.
class TokenException(SecureException):
    """Base class for secure token-related exceptions."""

class InvalidToken(SecureException):
    """A security check has failed on a token."""

class MissingToken(InvalidToken):
    """An expected token is missing."""
```

Comment (seanh) [reply src/h_vialib/exceptions.py]: The minimal change I think would be:

```python
# h_vialib/exceptions.py

class ViaLibException(Exception):
    """Base class for h_vialib exceptions."""

class InvalidToken(ViaLibException):
    """A security check has failed on a token."""

class MissingToken(ViaLibException):
    """An expected token is missing."""
```

Rather than the full version with two separate `exceptions.py` files. The `TokenException` base class in the middle would still be optional

Comment (jon-betts) [reply src/h_vialib/exceptions.py]: I don't think that's how exception handling works, otherwise every catch statement would trigger on raising `BaseException`. A `MissingToken` is an instance of `InvalidToken`, but an `InvalidToken` is not an instance of `MissingToken`.

The following code could do something like what you are suggesting:

```
try:
    ...
except InvalidToken:
    ...
except MissingToken:
    ...
```
But this is true of any inheritance in exceptions, including the fix suggested here:

```
try:
    ...
except TokenException:
    ...
except MissingToken:
    ...
```

Comment (seanh) [reply src/h_vialib/exceptions.py]: Yes, with a separate base class the order of exceptions still matters, but I'm saying that the separate base class makes the mistake easier to spot / harder to make. You always have to look at a list of `except`'s and make sure that they're in the right order: base classes must come last. With a separate base class like `TokenException` it's more obvious just from the name that it's a base class. Whereas with `MissingToken` and `InvalidToken` it's surprising that `InvalidToken` is actually a base class and must come last. You wouldn't spot it.

To put it another way: there's no reason to catch a `TokenException` base class _unless_ you want a catch-all `except` for all token-related errors

Comment (jon-betts) [reply src/h_vialib/exceptions.py]: I think the above is just a boo boo. @seanh explained in slack he does mean the case I'm suggesting here. Basically if the exceptions don't have an implied hierarchy in the name it makes it easier to accidentally catch calling a parent above a child.

I'm ok with introducing hierarchy for exceptions you want to catch (like this token one), but I don't think there's a reason to introduce a `ViaLibException`. Exceptions that come from `vialib` have no particular semantic meaning in the general case (unlike `requests` or something for example).

Comment (seanh) [reply src/h_vialib/exceptions.py]: Here's the reason why this should have a `VialibException` base class:

By adding a `TokenException` base class and having `InvalidToken` and `MissingToken` inherit from it you've captured all the advantages of our standard guide for exception classes (the `VialibException` approach) _in this particular case_ but to do that you had to spot the relation between `MissingToken` and `InvalidToken`, realise that they should be in an inheritance hierarchy and that there should be a third `TokenException` base class. There are lots of ways to get it wrong: `MissingToken` inheriting from `InvalidToken`, or neither of them inheriting from anything shared parent (except `Exception`), inheriting from Pyramid's exception classes, etc.

If we follow a standard guide that every package has a base `FoopackageException` that all its exceptions inherit from then, despite the semantic meaninglessness of `FoopackageException`, you get all the practical benefits without all the thinking and opportunities for creative ways to mess it up:

* You can catch `InvalidToken` and `MissingToken` separately, in any order
* If you want to catch both, you can catch `FoopackageException`
* If you come along later and add a new `ExpiredToken` class my code that catches `FoopackageException` will still work
* You can still add the intermediate `TokenException` base class if you want, but it's no longer necessary
* The guide also encourages exception classes to be defined not too near to and not too far from the code that raises them, to have not too many and not too few separate exception classes

Basically I think if everyone just follows the guide we'll be alright, whereas if everyone solves this problem anew for themselves we'll end up with an exciting variety of problems
